cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedSQLEngineCollection.java

```diff
@@ -166,7 +178,57 @@ public Runnable createMultiStoreTask(PhaseSpec phaseSpec,
                                        Set<String> group,
                                        Set<String> sinks,
                                        Map<String, StageStatisticsCollector> collectors) {
-    return unwrap().createMultiStoreTask(phaseSpec, group, sinks, collectors);
+    return new Runnable() {
```

Comment:
I see that in most cases we have business login in the wrapped? Can we do same here? It would ensure logic is not scattered

Reply:
I guess that makes sense, but we'll have to move the `storeTask` into the SQLEngineCollection as well to be consistent.

Reply:
@tivv I just walked through the code and tested a bit. Unfortunately, the logic to `unwrap` a wrapped collection requires a `pull` operation in order to route records to the right output port (which we use for Wrangler or other transforms with multiple potential outputs).. We need to keep this logic within the `wrappedSQLEngineCollection` class in order to prevent records from getting pulled. unless we need them to.

Reply:
It's not about `unwrap`, it's about using `wrapped`. We already do it for other store tasks

Reply:
For the `SQLEngineCollection` implementation, we perform a `pull` operation (which generates an `RDDCollection`) which we then use to perform the store tasks.

In this case, my goal is to validate if this step can be skipped before we go ahead and perform the `pull` operation.

I can try to refactor this to consolidate all logic in the `SQLEngineCollection` class, but I see one problem:

For the `multiStoreTask`, parameters can change based on what happens while the task is generated, This means I need yet another task to build this function based on arguments (the original group/sinks or the "trimmed" group/sinks without the tasks that were already processed). In my opinion this adds additional complexity to the overall code with not much value being added.

Let me know if you feel strongly about this and I'll go ahead and make these changes.

Reply:
Essentially, we would need to provide the "wrapped" collection a back reference into the original caller, in order to be able to follow the original path if this cannot be handled by this wrapped collection.

But since `WrappedSQLEngineCollection<T, U>` has 2 generic types and `SQLBackedCollection<U>` only has and needs one, this requires adding additional refactoring so the base `SQLEngineCollection` can correctly point back to the caller when needed.

Reply:
Does it make sense to have public method in the `SQLEngineCollection` class as `tryMultiStoreDirect` and then move all the logic from here to that class, so this Runnable would look something like

```java
public void run() {
  Pair<Set, Set> remaining = wraped.tryMultiStoreDirect(stageSpec);
  if (remaining != null) {
       unwrap().createMultiStoreTask(phaseSpec, pair.first, pair.second, collectors).run();
  } 
}
```